pkg/controller/stargazer.go

```diff
@@ -0,0 +1,247 @@
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	nimbessclientset "github.com/nimbess/stargazer/pkg/client/clientset/versioned"
+	unpinformer "github.com/nimbess/stargazer/pkg/client/informers/externalversions"
+	"github.com/nimbess/stargazer/pkg/config"
+	"github.com/nimbess/stargazer/pkg/etcdv3"
+	"github.com/nimbess/stargazer/pkg/handlers"
+	"github.com/nimbess/stargazer/pkg/signals"
+	"github.com/nimbess/stargazer/pkg/utils"
+	"reflect"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const maxRetries = 5
+
+var serverStartTime time.Time
+
+// Event indicate the informerEvent
+type Event struct {
+	key          string
+	eventType    string
+	namespace    string
+	resourceType string
+}
+
+// Controller object
+type Controller struct {
+	logger       *log.Entry
+	clientset    nimbessclientset.Interface
+	queue        workqueue.RateLimitingInterface
+	informer     cache.SharedIndexInformer
+	eventHandler handlers.Handler
+	etcdClient   etcdv3.Client
+}
+
+// Runs stargazer and then waits for process termination signals
+func Run(conf *config.Config, kubeClient *nimbessclientset.Clientset, etcdClient etcdv3.Client, ctx context.Context) {
+	v := reflect.ValueOf(conf.Controllers)
+	defer utilruntime.HandleCrash()
+	stopCh := signals.SetupSignalHandler()
+	ctrlType := v.Type()
+	for i := 0; i < v.NumField(); i++ {
+		if v.Field(i).Interface() == true {
+			log.Infof("Enabling controller for: %s", ctrlType.Field(i).Name)
+			// create handler
+			thisHandler, ok := handlers.Map[ctrlType.Field(i).Name]
+			if !ok {
+				log.Fatalf("Unsupported handler for controller: %s", ctrlType.Field(i).Name)
+			}
+			if err := thisHandler.Init(conf, etcdClient, ctx); err != nil {
+				log.Fatalf("Failed to init handler: %s", ctrlType.Field(i).Name)
+			}
+			c := Start(conf, kubeClient, thisHandler, stopCh)
+			defer c.queue.ShutDown()
+			log.Infof("Controller started: %s", ctrlType.Field(i).Name)
+		}
+	}
+
+	<-stopCh
+}
+
+// Start prepares a watcher and run corresponding controllers. Non-blocking. Returns new controller object.
+func Start(conf *config.Config, kubeClient *nimbessclientset.Clientset, eventHandler handlers.Handler,
+	stopCh <-chan struct{}) *Controller {
+
+	var informer cache.SharedIndexInformer
+	var resType string
+
+	nimbessInformerFactory := unpinformer.NewSharedInformerFactory(kubeClient, time.Second*30)
```

Comment:
does this return a new instance of the factory each time? It might be better to create a single factor once and then use it to spin up each controller informer. I believe I read that was the suggested model for better performance - to actually share the factory. Somehow it reduces the amount of communication.

Reply:
The factory var itself isn't significant, the real performance is about using a shared informer instead of a standalone one. The gain with using shared is on the backend k8s will only issue a single API call to update the cache for that resource, rather than 1 per subscriber. That is achieved by using a SharedInformerFactory here.